src/nsolid/nsolid_heap_snapshot.cc

```diff
@@ -350,5 +539,133 @@ void NSolidHeapSnapshot::snapshot_cb(uint64_t thread_id,
+
+  if (er) {
+    // In case the the thread is already gone, the cpu profile will be stopped
+    // on RemoveEnv, so do nothing here.
+  }
+}
+
+void NSolidHeapSnapshot::stop_sampling_profiler(
+    SharedEnvInst envinst,
+    uint64_t snaphot_id,
+    NSolidHeapSnapshot* snapshotter) {
+  uint64_t thread_id = envinst->thread_id();
+  nsuv::ns_mutex::scoped_lock lock(&snapshotter->in_progress_heap_sampling_);
+  auto it = snapshotter->threads_running_heap_sampling_.find(thread_id);
+  if (it == snapshotter->threads_running_heap_sampling_.end()) {
+    // The profiler was stopped before the timer was triggered
+    return;
+  }
+
+  HeapSamplerStor& stor = it->second;
+  if (stor.snapshot_id != snaphot_id) {
+    // The snapshot was stopped before the timer was triggered
+    return;
+  }
+
+  ASSERT(stor.flags & kFlagIsSamplingHeap);
+
+  // Give control back to the V8 thread
+  int er = RunCommand(envinst,
+                      CommandType::Interrupt,
+                      take_sampled_snapshot,
+                      snapshotter);
+  if (er) {
+    // In case the the thread is already gone, the cpu profile will be stopped
+    // on RemoveEnv, so do nothing here.
+  }
+}
+
+void NSolidHeapSnapshot::take_sampled_snapshot(
+    SharedEnvInst envinst,
+    NSolidHeapSnapshot* snapshotter) {
+  v8::Isolate* isolate = envinst->isolate();
+  v8::HeapProfiler* profiler = isolate->GetHeapProfiler();
+  ASSERT_NOT_NULL(profiler);
+
+  v8::HandleScope scope(isolate);
+
+  // Get the sampled heap profile
+  v8::AllocationProfile* sampled = profiler->GetAllocationProfile();
+
+  uint64_t thread_id = envinst->thread_id();
+  nsuv::ns_mutex::scoped_lock lock(&snapshotter->in_progress_heap_sampling_);
+  auto it = snapshotter->threads_running_heap_sampling_.find(thread_id);
+  ASSERT(it != snapshotter->threads_running_heap_sampling_.end());
+
+  HeapSamplerStor& stor = it->second;
+  // If this condition is reached. This was called by EnvList::RemoveEnv.
+  // It wants to stop any pending snapshot w/ tracking heap object.
+  if (!(stor.flags & kFlagIsSamplingHeap) || stor.flags & kFlagIsDone) {
+    // If no pending trackers, just do nothing
+    // There are not peding snapshots with trackers
+    return;
+  }
+
+  if (sampled == nullptr) {
+    QueueCallback(sampling_cb,
+                  thread_id,
+                  // TODO(juan) - this is not a heap snapshot failure
+                  heap_profiler::HEAP_SNAPSHOT_FAILURE,
+                  std::string(),
+                  snapshotter);
+  } else {
+    nlohmann::ordered_json profile_json =
+      build_v8_allocation_profile(isolate, sampled);
+    nlohmann::ordered_json samples_json;
+    // Send the whole profile to the callback
+    stor.cb(0, profile_json.dump().c_str(), stor.data.get());
+    // Send the empty string to signal the end of the profile
+    stor.cb(0, std::string(), stor.data.get());
+    // Stop the profiler
+    profiler->StopSamplingHeapProfiler();
```

Comment:
On this Sync version it's ok to call the cb from JS thread but not in the async one. Please. pass the profile to the N|Solid thread using QueueCallback(), in the same way you do above in the error path.

Reply:
Done